doc/lua_api.md

```diff
@@ -8898,7 +8898,10 @@ Player properties need to be saved manually.
 Entity definition
 -----------------
 
-Used by `minetest.register_entity`.
+Used by `minetest.register_entity`.  
+The entity definition table becomes a metatable of a newly created per-entity
+luaentity table, meaning its fields (e.g. `initial_properties`) will be shared
```

Comment:
I understand this has been merged, but the wording feels a bit confusing to me. When it says shared between all instances, this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.

As I understood it, the initial properties are used to set up a new instance of the entity (pass by value).

Wording such as: "The entity definition table is used in the creation of a metatable for a per-entity luaentity table, meaning its fields (e.g. `initial_properties`) will become the default values for each newly created instance of the entity." would be more clear.

If I have misunderstood, please let me know as I'm curious now.

Reply:
I just stumbled across the discussion after this, and felt Small Joker's wording was less ambigous:

"The entity definition table becomes a metatable of a newly created per-entity
luaentity table, roughly meaning its fields (e.g. `initial_properties`) will be
provided as default values."

Reply:
I think some of the discussion implies that modders should know Lua deeply, and I do not think that is the case for most people modding on minetest. They come in to it from different languages and different background. metatables aren't something known in many languages, but by reference and value is.

They want to make games and this could be the first place they learn lua.

Reply:
> this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.

this is correct.
what might be confusing you is that `initial_properties` is read by the engine once on object creation. any further property changes have nothing to do with this object, there isn't even a property table stored in Lua.
but if you edit the `initial_properties` table it will indeed affect all future instances.

Reply:
> > this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.
> 
> this is correct. what might be confusing you is that `initial_properties` is read by the engine once on object creation. any further property changes have nothing to do with this object, there isn't even a property table stored in Lua. but if you edit the `initial_properties` table it will indeed affect all future instances.

Interesting. Appreciate you taking the time to clarify this.

So when you set properties, this will have some form of override mechanism, or simply set up on the object (which was set on creation)?

Reply:
> When it says shared between all instances, this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.

This is the case. If you do `entity1.initial_properties.pointable = false` (which is not something you should do), then `entity2.initial_properties.pointable` will suddenly also be `false`, even if it was true before (if `entity1` and `entity2` are of the same type). This is because `entity1.initial_properties == entity2.initial_properties` which is just `minetest.registered_entities[entity1.name].initial_properties`.

> As I understood it, the initial properties are used to set up a new instance of the entity (pass by value).

Yes, on the C++ side of things, but that affects `self.object:[gs]et_properties`, not `self.initial_properties`.

---

Honestly at this point I think it'd probably be best to just specify exactly what we do, along the lines of:

> When Minetest instantiates an entity `self`, it sets the entity definition table `def` as the metatable ("prototype") of the object. The entity definition table has its `__index` field referring to itself. (This means that if the `self` table does not directly contain a key `k`, `self[k]` will evaluate to `def[k]`. For details refer to the Lua reference manual on metatables.)

But generally I don't think we should be explaining metatables *here* (if anything, the concept of prototypical OOP / metatables should be introduced as a term in a sort of "Lua primer" along with other terms which come up, and then be referenced here).

---

> I think some of the discussion implies that modders should know Lua deeply

Metatables are pretty fundamental, and existing resources like the PIL and the reference manual (and probably many other) already explain them well enough. They aren't that hard to understand. And similar prototypical OOP is found in e.g. JavaScript.

> but by reference and value is

Yes, but unfortunately not exactly applicable here. There is no way to be precise about what's going on here without speaking of metatables (at least the `__index` metamethod / field). This requires that the user knows what a metatable with `__index` set does.

Reply:
> > When it says shared between all instances, this reads to me that it is referenced by all instances (as in all instances access the same object). The implication being that if it is changed, it is changed for all.
> 
> This is the case. If you do `entity1.initial_properties.pointable = false` (which is not something you should do), then `entity2.initial_properties.pointable` will suddenly also be `false`, even if it was true before (if `entity1` and `entity2` are of the same type). This is because `entity1.initial_properties == entity2.initial_properties` which is just `minetest.registered_entities[entity1.name].initial_properties`.
> 
> > As I understood it, the initial properties are used to set up a new instance of the entity (pass by value).
> 
> Yes, on the C++ side of things, but that affects `self.object:[gs]et_properties`, not `self.initial_properties`.
> 
> Honestly at this point I think it'd probably be best to just specify exactly what we do, along the lines of:
> 

Thanks for clarifying

> > When Minetest instantiates an entity `self`, it sets the entity definition table `def` as the metatable ("prototype") of the object. The entity definition table has its `__index` field referring to itself. (This means that if the `self` table does not directly contain a key `k`, `self[k]` will evaluate to `def[k]`. For details refer to the Lua reference manual on metatables.)
> 

That's pretty clear imho.

> But generally I don't think we should be explaining metatables _here_ (if anything, the concept of prototypical OOP / metatables should be introduced as a term in a sort of "Lua primer" along with other terms which come up, and then be referenced here).
> 

This is a really good idea.

> > I think some of the discussion implies that modders should know Lua deeply
> 
> Metatables are pretty fundamental, and existing resources like the PIL and the reference manual (and probably many other) already explain them well enough. They aren't that hard to understand. And similar prototypical OOP is found in e.g. JavaScript.
> 

I'll probably shame myself in saying this, but I do not understand metatables well. I do think I would like to do some more reading. When I started reading about them, it did feel a bit confusing. I'm really confident in OOP in say Java, and languages like it, but when it comes to metatables, I think it unclear and confusing. Maybe I need to revisit it.

I've been a developer using OOP for over a decade and it isn't as easy as you imply. I think the reality is you are ridiculously intelligent and with the CS background, it makes things seem easy that probably are not. I really enjoy when you go into detail and explain stuff as I always learn things. I'm in a group of modders that uses a language to achieve things, and I'm usually successful in it. I enjoy learning more, but if learning more doesn't seem to have a tangible benefit to my ability to make things, I struggle with motivation. Long covid brain doesn't help that also, I guess.

I always find javascript hard to read. The syntax doesn't seem to work well with my brain. I can change it, but I'm not expressive in it, and it does take some effort.

> > but by reference and value is
> 
> Yes, but unfortunately not exactly applicable here. There is no way to be precise about what's going on here without speaking of metatables (at least the `__index` metamethod / field). This requires that the user knows what a metatable with `__index` set does.

I appreciate the clarification, and probably do need to dig into it further.

Thanks for your time and the response on this.

Reply:
> This is a really good idea.

I have opened a draft PR: https://github.com/minetest/minetest/pull/15112